src/_pytest/fixtures.py

```diff
@@ -1517,45 +1507,43 @@ def _getautousenames(self, nodeid: str) -> Iterator[str]:
         parentnode: nodes.Node,
+        initialnames: Tuple[str, ...],
+        arg2fixturedefs: Union[Dict[str, Sequence[FixtureDef[Any]]], None],
         ignore_args: Sequence[str] = (),
-    ) -> Tuple[Tuple[str, ...], List[str], Dict[str, Sequence[FixtureDef[Any]]]]:
+    ) -> Tuple[List[str], Dict[str, Sequence[FixtureDef[Any]]]]:
         # Collect the closure of all fixtures, starting with the given
-        # fixturenames as the initial set.  As we have to visit all
-        # factory definitions anyway, we also return an arg2fixturedefs
-        # mapping so that the caller can reuse it and does not have
-        # to re-discover fixturedefs again for each fixturename
+        # initialnames containing function arguments, `usefixture` markers
+        # and `autouse` fixtures as the initial set.  As we have to visit all
+        # factory definitions anyway, we also populate arg2fixturedefs mapping
+        # for the args missing therein so that the caller can reuse it and does
+        # not have to re-discover fixturedefs again for each fixturename
         # (discovering matching fixtures for a given name/node is expensive).
 
-        parentid = parentnode.nodeid
-        fixturenames_closure = list(self._getautousenames(parentid))
-
-        def merge(otherlist: Iterable[str]) -> None:
-            for arg in otherlist:
-                if arg not in fixturenames_closure:
-                    fixturenames_closure.append(arg)
-
-        merge(fixturenames)
+        fixturenames_closure = initialnames
 
-        # At this point, fixturenames_closure contains what we call "initialnames",
-        # which is a set of fixturenames the function immediately requests. We
-        # need to return it as well, so save this.
-        initialnames = tuple(fixturenames_closure)
-
-        arg2fixturedefs: Dict[str, Sequence[FixtureDef[Any]]] = {}
+        if arg2fixturedefs is None:
+            arg2fixturedefs = {}
         lastlen = -1
+        parentid = parentnode.nodeid
         while lastlen != len(fixturenames_closure):
             lastlen = len(fixturenames_closure)
             for argname in fixturenames_closure:
                 if argname in ignore_args:
                     continue
-                if argname in arg2fixturedefs:
-                    continue
-                fixturedefs = self.getfixturedefs(argname, parentid)
-                if fixturedefs:
-                    arg2fixturedefs[argname] = fixturedefs
-                    merge(fixturedefs[-1].argnames)
+                if argname not in arg2fixturedefs:
+                    fixturedefs = self.getfixturedefs(argname, parentid)
+                    if fixturedefs:
+                        arg2fixturedefs[argname] = fixturedefs
+                else:
+                    fixturedefs = arg2fixturedefs[argname]
+                if fixturedefs and not (
+                    fixturedefs[-1].func.__name__ == "get_direct_param_fixture_func"
```

Comment:
This is something that IMO we should avoid, too magic and will be missed on a refactoring...

Reply:
@nicoddemus , sorry for delay. As I stated in the PR description, I need `is_pseudo` to get rid of this untidy expression. I added `IdentityFixture` as we discussed in [PR#11220](https://github.com/pytest-dev/pytest/pull/11220#discussion_r1277288411) and replaced this expression with an instance check. How is this?